### PR TITLE
[SPARK-10367][SQL] Support Parquet logical type INTERVAL

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?

Adds support for writing and reading interval data into parquet files.
In this PR, we encode Catalyst CalendarIntervalType to Parquet's FIXED_LEN_BINARY
annotated by INTERVAL. This is in WIP state for further discussions.
## How was this patch tested?

Added new tests in ParquetIOSuite

### CODE REVIEW

**Comment by dilipbiswal** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 1]
```diff
@@ -672,6 +672,8 @@ object ScalaReflection extends ScalaReflection {
       case t if t <:< localTypeOf[java.sql.Timestamp] => Schema(TimestampType, nullable = true)
       case t if t <:< localTypeOf[java.sql.Date] => Schema(DateType, nullable = true)
       case t if t <:< localTypeOf[BigDecimal] => Schema(DecimalType.SYSTEM_DEFAULT, nullable = true)
+      case t if t <:<  localTypeOf[CalendarInterval] =>
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: Remove the additional space here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 2]
```diff
@@ -435,7 +435,9 @@ case class DataSource(
   def write(
       mode: SaveMode,
       data: DataFrame): BaseRelation = {
-    if (data.schema.map(_.dataType).exists(_.isInstanceOf[CalendarIntervalType])) {
+
+    if (data.schema.map(_.dataType).exists(_.isInstanceOf[CalendarIntervalType]) &&
+      className != "parquet") {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: className.toLowerCase.

**Comment by gatorsmile** [comment]
Saw @viirya submitted a PR for the same issue: https://github.com/apache/spark/pull/7793

**Comment by dilipbiswal** [comment]
@gatorsmile Thanks .. I had absolutely no idea about this PR :-)

**Comment by gatorsmile** [comment]
retest this please

**Review by gatorsmile — COMMENTED** [review]
Basically, this PR does not enable us to write the data of interval type to a parquet file, but it allows us to read the parquet files containing the interval types. Do you know any system that write such parquet files with interval types? Based on this factor, we can consider whether we should support it. Thanks!

**Comment by dilipbiswal** [comment]
@gatorsmile Hi Sean, i am sorry, i didn't see your ping on this pr. I will get back to you on your question as i need to re-create my env. I do remember one thing though. I do remember looking for writers to test this newly added logical type. But i didn't find any.. and i had to create the data files using parquet apis. I will re-check again and get back.

Since we are coding here to a defined parquet spec, i felt safe to add this PR. We know that there are databases such as postgress, oracle , informix which support interval type. So if we were to make this data available to spark to read, they would have the option to encode their interval data in parquet more naturally and spark would be able to read it.

**Comment by gatorsmile** [comment]
If we can find the other writer that can generate such a parquet file, we should support it; otherwise, we will hold it until we find such writers. Thank you!

**Comment by dilipbiswal** [comment]
@gatorsmile Hi Sean, i tried apache-drill after looking through their documentation. And they are able to encode interval data into parquet.

```
0: jdbc:drill:zk=local> CREATE TABLE dfs.tmp.parquet_intervals AS 
. . . . . . . . . . . > (SELECT CAST( INTERVALYEAR_col as INTERVAL YEAR) INTERVALYEAR_col,
. . . . . . . . . . . > CAST( INTERVALDAY_col as INTERVAL DAY) INTERVALDAY_col,
. . . . . . . . . . . > CAST( INTERVAL_col as INTERVAL SECOND) INTERVAL_col 
. . . . . . . . . . . > FROM dfs.`/tmp/intervals.json`);
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
+-----------+----------------------------+
| Fragment  | Number of records written  |
+-----------+----------------------------+
| 0_0       | 3                          |
+-----------+----------------------------+
```
Here is the schema of the written parquet file.
```
message root {
  optional fixed_len_byte_array(12) INTERVALYEAR_col (INTERVAL);
  optional fixed_len_byte_array(12) INTERVALDAY_col (INTERVAL);
  optional fixed_len_byte_array(12) INTERVAL_col (INTERVAL);
}

```
From presto's documentation, it seems like they also may be able to encode interval data. But i haven't tried. 

FYI - i also tried hive. Its not possible to encode interval data in parquet format through hive.

**Comment by maropu** [comment]
IIUC we have no plan to expose Interval types now, so can we close this for now? cc: @gatorsmile

**Comment by maropu** [comment]
oh, I noticed the jira ticket has already been closed as later, so can you close this? @dilipbiswal

**Comment by dilipbiswal** [comment]
@maropu Thanks.. closing it.

**Pull request closed without merging**